File: src/main/kotlin/timmybot/AwsSecretsService.kt
package timmybot

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.KotlinModule
import mu.KotlinLogging
import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueRequest

/**
 * Service for retrieving secrets from AWS Secrets Manager
 */
class AwsSecretsService {
    
    private val logger = KotlinLogging.logger {}
    private val secretsClient: SecretsManagerClient
    private val objectMapper = ObjectMapper().registerModule(KotlinModule.Builder().build())
    
    init {
        val region = System.getenv("AWS_DEFAULT_REGION") ?: System.getenv("AWS_REGION") ?: "eu-central-1"
        secretsClient = SecretsManagerClient.builder()
            .region(Region.of(region))
            .credentialsProvider(DefaultCredentialsProvider.create())
            .build()
            
        logger.info { "AwsSecretsService initialized for region: $region" }
    }
    
    /**
     * Get Discord bot token from AWS Secrets Manager
     */
    fun getDiscordBotToken(): String {
        val secretName = System.getenv("DISCORD_BOT_TOKEN_SECRET") ?: "timmybot/dev/discord-bot-token"
        
        try {
            val request = GetSecretValueRequest.builder()
                .secretId(secretName)
                .build()
                
            val response = secretsClient.getSecretValue(request)
            val secretString = response.secretString()
            
            // Parse the JSON secret
            val secretMap = objectMapper.readValue(secretString, Map::class.java) as Map<String, Any>
            val token = secretMap["token"] as? String
            
            if (token.isNullOrBlank()) {
                throw IllegalStateException("Discord bot token is empty in secret")
            }
            
            logger.info { "Successfully retrieved Discord bot configuration from Secrets Manager" }
            return token
            
        } catch (e: Exception) {
            logger.error(e) { "Failed to retrieve Discord bot token from secret: $secretName" }
            throw IllegalStateException("Could not retrieve Discord bot token from Secrets Manager. Check AWS credentials and permissions.", e)
        }
    }
    
    /**
     * Get database configuration from AWS Secrets Manager
     */
    fun getDatabaseConfig(): DatabaseConfig {
        val secretName = System.getenv("DATABASE_CONFIG_SECRET") ?: "timmybot/dev/database-config"
        
        try {
            val request = GetSecretValueRequest.builder()
                .secretId(secretName)
                .build()
                
            val response = secretsClient.getSecretValue(request)
            val secretString = response.secretString()
            
            // Parse the JSON secret
            val secretMap = objectMapper.readValue(secretString, Map::class.java) as Map<String, Any>
            
            return DatabaseConfig(
                region = secretMap["dynamodb_region"] as? String ?: "eu-central-1",
                guildQueuesTable = secretMap["guild_queues_table"] as? String ?: "timmybot-dev-guild-queues",
                userPreferencesTable = secretMap["user_preferences_table"] as? String ?: "timmybot-dev-user-prefs",
                trackCacheTable = secretMap["track_cache_table"] as? String ?: "timmybot-dev-track-cache",
                serverAllowlistTable = secretMap["server_allowlist_table"] as? String ?: "timmybot-dev-server-allowlist"
            )
            
        } catch (e: Exception) {
            logger.error(e) { "Failed to retrieve database config from secret: $secretName" }
            // Fallback to environment variables
            return DatabaseConfig(
                region = System.getenv("AWS_REGION") ?: "eu-central-1",
                guildQueuesTable = System.getenv("GUILD_QUEUES_TABLE") ?: "timmybot-dev-guild-queues",
                userPreferencesTable = System.getenv("USER_PREFERENCES_TABLE") ?: "timmybot-dev-user-prefs",
                trackCacheTable = System.getenv("TRACK_CACHE_TABLE") ?: "timmybot-dev-track-cache",
                serverAllowlistTable = System.getenv("SERVER_ALLOWLIST_TABLE") ?: "timmybot-dev-server-allowlist"
            )
        }
    }
    
    /**
     * Get application configuration from AWS Secrets Manager
     */
    fun getAppConfig(): AppConfig {
        val secretName = System.getenv("APP_CONFIG_SECRET") ?: "timmybot/dev/app-config"
        
        try {
            val request = GetSecretValueRequest.builder()
                .secretId(secretName)
                .build()
                
            val response = secretsClient.getSecretValue(request)
            val secretString = response.secretString()
            
            // Parse the JSON secret
            val secretMap = objectMapper.readValue(secretString, Map::class.java) as Map<String, Any>
            
            return AppConfig(
                environment = secretMap["environment"] as? String ?: "dev",
                logLevel = secretMap["log_level"] as? String ?: "INFO",
                serverAllowlistEnabled = (secretMap["server_allowlist_enabled"] as? String)?.toBoolean() ?: true,
                oauthRequired = (secretMap["oauth_required"] as? String)?.toBoolean() ?: false,
                premiumFeatures = (secretMap["premium_features"] as? String)?.toBoolean() ?: true
            )
            
        } catch (e: Exception) {
            logger.error(e) { "Failed to retrieve app config from secret: $secretName" }
            // Fallback to environment variables
            return AppConfig(
                environment = System.getenv("ENVIRONMENT") ?: "dev",
                logLevel = System.getenv("LOG_LEVEL") ?: "INFO",
                serverAllowlistEnabled = System.getenv("SERVER_ALLOWLIST_ENABLED")?.toBoolean() ?: true,
                oauthRequired = System.getenv("OAUTH_REQUIRED")?.toBoolean() ?: false,
                premiumFeatures = System.getenv("PREMIUM_FEATURES")?.toBoolean() ?: true
            )
        }
    }
    
    data class DatabaseConfig(
        val region: String,
        val guildQueuesTable: String,
        val userPreferencesTable: String,
        val trackCacheTable: String,
        val serverAllowlistTable: String
    )
    
    data class AppConfig(
        val environment: String,
        val logLevel: String,
        val serverAllowlistEnabled: Boolean,
        val oauthRequired: Boolean,
        val premiumFeatures: Boolean
    )
}

File: discord-js/src/services/AwsSecretsService.ts
import { SecretsManagerClient, GetSecretValueCommand } from '@aws-sdk/client-secrets-manager';
import { logger } from '../utils/logger';

/**
 * Configuration interfaces matching the Kotlin implementation
 */
export interface DatabaseConfig {
  region: string;
  guildQueuesTable: string;
  userPreferencesTable: string;
  trackCacheTable: string;
  serverAllowlistTable: string;
}

export interface AppConfig {
  environment: string;
  logLevel: string;
  serverAllowlistEnabled: boolean;
  oauthRequired: boolean;
  premiumFeatures: boolean;
}

/**
 * Service for retrieving secrets from AWS Secrets Manager
 * Maintains exact same interface as existing Kotlin implementation for compatibility
 */
export class AwsSecretsService {
  private secretsClient: SecretsManagerClient;
  private readonly maxRetries = 3;
  private readonly baseDelayMs = 1000;

  constructor() {
    const region = process.env.AWS_DEFAULT_REGION || process.env.AWS_REGION || 'eu-central-1';
    
    this.secretsClient = new SecretsManagerClient({
      region,
      maxAttempts: this.maxRetries,
    });

    logger.info(`AwsSecretsService initialized for region: ${region}`);
  }

  /**
   * Get Discord bot token from AWS Secrets Manager
   */
  async getDiscordBotToken(): Promise<string> {
    const secretName = process.env.DISCORD_BOT_TOKEN_SECRET || 'timmybot/dev/discord-bot-token';

    try {
      const secretString = await this.getSecretWithRetry(secretName);
      
      // Parse the JSON secret with proper error handling
      let secretMap: Record<string, any>;
      try {
        secretMap = JSON.parse(secretString);
      } catch (error) {
        logger.error(`Failed to parse secret JSON: ${secretName}`, error);
        throw new Error('Invalid secret JSON format in Secrets Manager');
      }

      if (typeof secretMap !== 'object' || secretMap === null) {
        logger.error(`Failed to cast secret to object format: ${secretName}`);
        throw new Error('Invalid secret format in Secrets Manager: expected JSON object');
      }

      const token = secretMap.token;

      if (!token || typeof token !== 'string' || token.trim() === '') {
        throw new Error('Discord bot token is empty in secret');
      }

      logger.info('Successfully retrieved Discord bot configuration from Secrets Manager');
      return token;

    } catch (error) {
      if (error instanceof Error && (
        error.message.includes('Discord bot token is empty') ||
        error.message.includes('Invalid secret JSON format') ||
        error.message.includes('Invalid secret format')
      )) {
        throw error;
      }
      logger.error(`Failed to retrieve Discord bot token from secret: ${secretName}`, error);
      throw new Error('Could not retrieve Discord bot token from Secrets Manager. Check AWS credentials and permissions.');
    }
  }

  /**
   * Get database configuration from AWS Secrets Manager
   */
  async getDatabaseConfig(): Promise<DatabaseConfig> {
    const secretName = process.env.DATABASE_CONFIG_SECRET || 'timmybot/dev/database-config';

    try {
      const secretString = await this.getSecretWithRetry(secretName);
      
      // Parse the JSON secret with proper error handling
      let secretMap: Record<string, any>;
      try {
        secretMap = JSON.parse(secretString);
      } catch (error) {
        logger.error(`Failed to parse secret JSON: ${secretName}`, error);
        throw new Error('Invalid secret JSON format in Secrets Manager');
      }

      if (typeof secretMap !== 'object' || secretMap === null) {
        logger.error(`Failed to cast secret to object format: ${secretName}`);
        throw new Error('Invalid secret format in Secrets Manager: expected JSON object');
      }

      return {
        region: secretMap.dynamodb_region || 'eu-central-1',
        guildQueuesTable: secretMap.guild_queues_table || 'timmybot-dev-guild-queues',
        userPreferencesTable: secretMap.user_preferences_table || 'timmybot-dev-user-prefs',
        trackCacheTable: secretMap.track_cache_table || 'timmybot-dev-track-cache',
        serverAllowlistTable: secretMap.server_allowlist_table || 'timmybot-dev-server-allowlist',
      };

    } catch (error) {
      if (error instanceof Error && error.message.includes('Invalid secret')) {
        throw error;
      }
      logger.error(`Failed to retrieve database config from secret: ${secretName}`, error);
      throw new Error('Could not retrieve database configuration from Secrets Manager. This is a production deployment - Secrets Manager is required.');
    }
  }

  /**
   * Get application configuration from AWS Secrets Manager
   */
  async getAppConfig(): Promise<AppConfig> {
    const secretName = process.env.APP_CONFIG_SECRET || 'timmybot/dev/app-config';

    try {
      const secretString = await this.getSecretWithRetry(secretName);
      
      // Parse the JSON secret with proper error handling
      let secretMap: Record<string, any>;
      try {
        secretMap = JSON.parse(secretString);
      } catch (error) {
        logger.error(`Failed to parse secret JSON: ${secretName}`, error);
        throw new Error('Invalid secret JSON format in Secrets Manager');
      }

      if (typeof secretMap !== 'object' || secretMap === null) {
        logger.error(`Failed to cast secret to object format: ${secretName}`);
        throw new Error('Invalid secret format in Secrets Manager: expected JSON object');
      }

      return {
        environment: secretMap.environment || 'dev',
        logLevel: secretMap.log_level || 'INFO',
        serverAllowlistEnabled: this.parseBoolean(secretMap.server_allowlist_enabled, true),
        oauthRequired: this.parseBoolean(secretMap.oauth_required, false),
        premiumFeatures: this.parseBoolean(secretMap.premium_features, true),
      };

    } catch (error) {
      if (error instanceof Error && error.message.includes('Invalid secret')) {
        throw error;
      }
      logger.error(`Failed to retrieve app config from secret: ${secretName}`, error);
      throw new Error('Could not retrieve application configuration from Secrets Manager. This is a production deployment - Secrets Manager is required.');
    }
  }

  /**
   * Get secret value with exponential backoff retry logic
   */
  private async getSecretWithRetry(secretName: string): Promise<string> {
    let lastError: Error | null = null;

    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      try {
        const command = new GetSecretValueCommand({
          SecretId: secretName,
        });

        const response = await this.secretsClient.send(command);
        
        if (!response.SecretString) {
          throw new Error('Secret value is empty');
        }

        return response.SecretString;

      } catch (error) {
        lastError = error as Error;
        
        if (attempt === this.maxRetries) {
          break;
        }

        // Calculate exponential backoff delay with jitter
        const delay = this.baseDelayMs * Math.pow(2, attempt - 1) + Math.random() * 1000;
        logger.warn(`Attempt ${attempt} failed for secret ${secretName}, retrying in ${Math.round(delay)}ms`, error);
        
        await this.sleep(delay);
      }
    }

    throw lastError || new Error(`Failed to retrieve secret after ${this.maxRetries} attempts`);
  }

  /**
   * Parse boolean values from string or boolean
   */
  private parseBoolean(value: any, defaultValue: boolean): boolean {
    if (typeof value === 'boolean') {
      return value;
    }
    if (typeof value === 'string') {
      const lowerValue = value.toLowerCase();
      if (lowerValue === 'true') return true;
      if (lowerValue === 'false') return false;
      // For invalid string values, return default
      return defaultValue;
    }
    return defaultValue;
  }

  /**
   * Sleep utility for retry delays
   */
  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}